core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreAggregateCaseAggregations.java

```diff
@@ -542,6 +620,50 @@ public int hashCode()
         }
     }
 
+    // Two CaseAggregations are mergeable if they satisfy the below conditions:
+    // 1. They have the same aggregation function
+    // 2. They have an IF expression in the result expression
+    // 3. The IF expressions have the same true and false values
+    public static class MergeabilityKey
```

Comment:
Why we need `MergeabilityKey` and `PreAggregationKey` is not sufficient?

Reply:
I don't think my convolutional code change is ideal, but cannot think of a better way. Initially I was trying an ad-hoc approach - 1. creating preaggregations as we go ; 2. when seeing a "similar to existing" (IF) case aggregation, create a merged preaggregation; 3. also update existing preaggregations with the merged preagg. I found this to be problematic because we cannot easily locate the existing "similar" entries and it's likely an O(n^2) search. Furthermore, there can be multiple iterations of merges to the entries (think of Q66) which are error prone.

So I dropped above idea and took this cleaner (hopefully) approach. Now we only need a pre-processing step to group the case aggregations based on their "IF-mergeability" when needed, and then create merged PreAggregationKey. That way in the normal loop, we can just assign each case aggregation to either a simple PreAggregationKey or a merged PreAggregationKey, knowing the projection part is final.

One key info we need in determining the "IF-mergeability" is, given two case aggregations - 1. whether they both have an IF expression in the result section (this can be either original IF, or an artificial one created by this PR); 2. and if so, whether they have the same true/false expressions. We intentionally want to skip the comparison of the IF condition part because that's what we hope to merge. The existing PreAggregationKey cannot fulfill this purpose, so I have to create the separate MergeabilityKey.

Reply:
>  found this to be problematic because we cannot easily locate the existing "similar" entries and it's likely an O(n^2) search.

Similar entries should be the ones with same `PreAggregationKey`, because `PreAggregationKey#projection` will `uniquely` determine if pre-projection require `IF` or not. Take for example,
```
max(CASE WHEN d_mon=1 THEN colA * colB end)
```
For every `PreAggregationKey` that has `PreAggregationKey#projection=colA * colB` a pre projection of `IF ... THEN colA * colB` needs to be created.

Hence. You can collect `CaseAggregations` with same `PreAggregationKey` and then build pre-projection, e.g:
```
Multimap<PreAggregateKey, CaseAggregation> groupedAggregations = aggregations.stream()
  collect(toImmutableListMultimap(PreAggregationKey::new, identity();

// HERE: construct PreAggregation from list of CaseAggregation for each PreAggregateKey
// e.g. pre-projection should be IF (aggregation1.operand OR aggregation2.operator ..., aggregation1.result)
return groupedAggregations.asMap().entries().stream()
  .collect(identity(), entry -> CONSTRUCT PreAggregation here)
```

Reply:
Here's my point on why the ad-hoc approach won't work well. Suppose we see two case expressions first:
```
max(CASE WHEN d_mon=1 THEN IF(d_mon=1, colA * colB end)) -- case1
max(CASE WHEN d_mon=2 THEN IF(d_mon=2, colA * colB end)) -- case2
```
We put into our map with the first case expression:
```
<max, IF(d_mon=1, colA * colB)> -> case1
```
When we see the second case expression, since its `PreAggregationKey` looks like `<max, IF(d_mon=2, colA * colB)>`, we won't be able to know "case2" is supposed to be merged with "case1". So the point is again about `PreAggregationKey` not able to provide the needed "identify" for the mergeable case expressions.

Reply:
> We put into our map with the first case expression:

We can just put `<max, colA * colB> -> (case1, case2)` in the map fist. Then we can create:
```
<max, colA * colB> -> PreAggregation(..., projection:= IF(d_mon=1 OR d_mon=2, colA * colB))
```

The reason I would like to keep the changes minimal is because the code here is already pretty complex

Reply:
The problem is when there is a third case expression, it's not straightforward to figure out whether this is to be merged with an existing entry in the map. And we have to recreate the PreAggregation. That's why I think a preprocessing step works better and is clearer.
```
max(CASE WHEN d_mon=3 THEN IF(d_mon=3, colA * colB end)) -- case3
```

Reply:
>  it's not straightforward to figure out whether this is to be merged with an existing entry in the map

why it's not straightforward? The rule is:
* merge all pre-projections with same `PreAggregationKey`.

>  And we have to recreate the PreAggregation. That's why I think a preprocessing step works better and is clearer.

We don't really have to re-create it, see https://github.com/trinodb/trino/pull/20652/commits/8b00d76eddbf0d4f1e93298401b9d1e6aebb919e#r1504884938